### PR TITLE
Implement Minimax Music 3 + Core Support for Cuda Graphs

### PR DESCRIPTION
**Pull request by rattus128**
This PR contains a new comfy model and a supporting major core feature (cuda graphs)

# Minimax Music 3 native model support

https://huggingface.co/Comfy-Org/MiniMax-Music-3

This is a two phase model with a non-trivial auto-regressive model feeding a smaller diffusion model.

Implement core support in comfy. A text encoder node with split fields for lyrics and caption and generation settings is added. The AR model steps are proportional to the generated output and are capable of detecting short end of output therefore the AR drives latent size generation.

image

### Model formats and quants

The original published model segmented the embedding and lm_head  weights into two distinct regions for text and audio without splitting the weights. This lead to a case where the giant embedding and lm_head weights were only being partially used or in the case of lm_head padding the checkpoint with dead weight and dead computation. Comfy has repacked the models to remove these dead rows on lm_head (-1.6GB).

The embedding has been split between the audio and text sections as the larger text section only needs to exist for the prefill stage of AR and should be freed from VRAM once you get the VRAM critical AR. So that is split and a new facility is added to dynamic VRAM to push that prefill only embedding down to the lowest priority and its first to go once you hit the VRAM ceiling.

for the AR, bf16 and int8_convrot are tested with good results
w4a8 has been briefly tested as runnable but with variable results (not shipped)
for DiT, fp32 (original) fp16 and int8_convrot are tested with good results

Rerformace of the AR models is often CPU bottlenecked in comfy, so now that we have this high-step-count AR model performance is unlocked with ...

# Cuda Graphs

Add native cuda graph support to ComfyUI interoperable with dynamic VRAM. The way it works, is a model declares its units of prefetch to the dynamic vram engine such that these units are consolidated in the aimdo VBAR. The existing block prefetcher is enabled for the model and the prefetch iterator declares the executable core for the block. The prefetch engine then aimdo faults the VRAM, and it its present in memory does nessacary one-time-prep (e.g. Lora application). The layer is then cuda graphed for replay. Each step the prefetched then only has to check if the vbar is still faulted and if so, replay the graph.

The graph implementation is kept simple, with a side-effect only policy on graph IO. That is, there are no pure input and outputs to the graph by design as cuda-graphs needs static addresses for memory allocations. The models that use this engine are therefore refactored to move the critical pieces of transformer block IO to static buffers that live for the duration of inferences. This is called "cross-step-state". A registry for this is added to model_management.

The most obvious piece of data that goes in cross-step-state is X (the activation that flows from one transformer block to the next). A single x buffer is allocated and then exists for the lifetime of inference and each graph reads from and write to this singular buffer. This avoids D2D copies and per graph allocations for IO buffers (which cost a lot of VRAM too).

The main code pattern to use this is this (from llama in this PR):

```
       for i, layer in enumerate(self.layers):
...
            def core():
                _, current_kv = layer(
                    x=x,
                    attention_mask=mask,
                    freqs_cis=freqs_cis,
                    optimized_attention=optimized_attention,
                    past_key_value=past_kv,
                )
                ...

            comfy.model_prefetch.prefetch_queue_pop(prefetch_queue, x.device, layer, x.dtype, core=core, enable_graph=enable_graph)

```

Graphs can be disabled with `--disable-cuda-graphs` at comfy startup.

### LLama and KV Cache Attention

The llama model as used by minimax uses a growing KV cache. The original code does cpu management of the constantly growing-by-one KV cache artifacts however the logical CPU control is incompatible with cuda graphs. This requires a sequence length controllable attention.

Comfy-kitchen v 0.2.31 introduces a minimal generic flash attention for bf16 that implments this. The attention length are GPU controlled buffers and their increments are single element tensor ops that are just part of the graph. The fully migrates the geometry control to the GPU and gets the CPU out of the way enabling cuda graphs.

### Minimax AR use of Graphs and layer priorities

Minimax AR is a demanding model for AR generation on two levels. The main transformer is a high-layer-rate model in its own right so this is graphed on a block-by-block basis. There is also and audio processing transfomer that runs 7 times per main AR step (so ultimately these layers are 7x the step state).

The audio transformer is therefore place at the ultimate priority for the dynamic VRAM engine to load (the new api for declaring prefetch blocks also lets you declare priority). This transformer is prefetched and graphed monolitically as its only 2GB of weights but maximizing layer-rate is critical through this section.

# Test results

120s generation (long prompt to overflow AR) https://github.com/Comfy-Org/ComfyUI/pull/15570#top

`minimax_music3_text_encoder_int8_convrot_pruned.safetensors`
`minimax_music3_dit_int8_convrot.safetensors`
`minimax_music3_dav.safetensors`

```
System                              CUDA graphs    Total       AR      DiT      DAV
----------------------------------  -----------  -------  -------  -------  -------
RTX 5090 (96 GB RAM) — Linux        enabled        87.7s    39.6s    46.0s     2.1s
RTX 5090 (96 GB RAM) — Linux        disabled      287.5s   238.7s    45.6s     2.0s
RTX 5090 (96 GB RAM) — Windows      enabled       137.0s    45.3s    86.8s     3.3s
RTX 5090 (96 GB RAM) — Windows      disabled      562.0s   471.2s    86.1s     3.2s
RTX 3060 (48 GB RAM) — Windows      enabled       538.8s   142.7s   143.6s   251.9s
RTX 3060 (48 GB RAM) — Windows      disabled      700.6s   307.1s   143.0s   249.8s
RTX 5060 (48 GB RAM) — Windows      enabled      1096.2s   913.3s   110.1s    71.8s
RTX 5060 (48 GB RAM) — Windows      disabled     1090.1s   921.1s    99.4s    68.5s
```

NOTE: the RTX5090 system is also underclocked to 2GHz to emulate slower CPU.

_NOTE: The DAV performance on 3060 is under investigation_

Preformance regression tests

```
Platform  Workload   Current  Baseline  Change
--------  ---------  -------  --------  ------
Linux     Z-Image      2.75s     2.78s   -1.0%
Linux     Ideogram    10.03s     9.68s   +3.6%
Windows   Z-Image      2.88s     2.98s   -3.4%
Windows   Ideogram    14.62s    14.71s   -0.6%
```

General regression tests:

Linux, RTX5090, 96GB, 2GHz underclock

```
- MiniMax H3 — completed in 79.09s ✅ 
- WAN 2.2 — completed in 51.57s ✅ 
- Stable Cascade → Flux2 — completed in 69.81s ✅ 
- Ace step turbo 1.5 XL ✅ 
- LTX2.3 ✅ 
```

All outputs human checked ✅

### CODE REVIEW

File: comfy/ops.py [hunk 1]
```diff
@@ -123,10 +123,12 @@ def materialize_meta_param(s, param_keys):
 
 
 # FIXME: add n=1 cache hit fast path
-def cast_modules_with_vbar(comfy_modules, dtype, device, bias_dtype, non_blocking):
+def cast_modules_with_vbar(comfy_modules, dtype, device, bias_dtype, non_blocking, return_faulted=False):
     offload_stream = None
     cast_buffer = None
     cast_buffer_offset = 0
+    if return_faulted:
+        fully_faulted = all(not getattr(s, param_key + "_function", []) for s in comfy_modules for param_key in ("weight", "bias"))
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/ops.py:131]: this is the truth. if there is nothing to fault theres nothing to do which means you are fully faulted. Your proposal is regressive as it would stop cuda graphs of an only-force-loaded or weightless module.

File: comfy/ldm/minimax_music/ar.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                depth_io["codes"].copy_(codes)
+                depth_io["depth_hidden"].copy_(depth_hidden)
+
+            depth_queue = comfy.model_prefetch.make_prefetch_queue(
+                [(decoder, self.model.audio_extra_embedding)], device, {"prefetch_dynamic_vbars": True}
+            )
+            comfy.model_prefetch.prefetch_queue_pop(
+                depth_queue, device, decoder, execution_dtype, core=depth_core, enable_graph=True, generator=generator
+            )
+            comfy.model_prefetch.prefetch_queue_pop(depth_queue, device, None)
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/ldm/minimax_music/ar.py:317]: its not the hard memory stuff that coding guildline is talking about. The prefetch queue is single use by design and all existing users are once per invocation. This is fine. Please elaborate where you think there is a leak or early return? cross-step-state explictly has a per node garbage collector for this reason.

File: comfy/ldm/minimax_music/dav.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    def forward(self, x):
+        return snake(x, self.alpha)
+
+
+def _weight_norm_conv(operations, *args, **kwargs):
+    return nn.utils.weight_norm(operations.Conv1d(*args, **kwargs))
+
+
+def _weight_norm_conv_transpose(operations, *args, **kwargs):
+    return nn.utils.weight_norm(operations.ConvTranspose1d(*args, **kwargs))
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/ldm/minimax_music/dav.py]: my agent says its just a case of updating to paramtrizations API variant

File: comfy/ldm/minimax_music/prompt.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    lines = []
+    for line in lyrics.split("\n"):
+        match = _LEADING_TAGS_RE.match(line)
+        lines.append(match.group(1).strip() if match else line)
+    text = "\n".join(lines)
+    text = text.replace("] ", "]\n")
+    text = text.replace(" [", "\n[")
+    text = text.replace(" ^ ", "\n")
+    text = re.sub(r"\[([^\]]+)\]", lambda match: f"[{match.group(1).lower()}]", text)
+    return f"[start]\n{text}"
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/ldm/minimax_music/prompt.py:53]: that whole loop looks bogus, and its a case of fixing the newline regexes below IIUC

File: comfy_extras/nodes_minimax_music.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                io.Int.Input("seed", default=0, min=0, max=0xffffffffffffffff, control_after_generate=True),
+                io.Float.Input("max_duration", default=120.0, min=0.04, max=MAX_AUDIO_FRAMES / 25, step=0.04, tooltip="Maximum duration in seconds; the model can end the song earlier."),
+                io.Float.Input("cfg_scale", default=CFG_SCALE, min=0.0, max=100.0, step=0.1, round=0.01, advanced=True),
+                io.Int.Input("top_k", default=CFG_TOP_K, min=1, max=C0_VOCAB_SIZE, advanced=True),
+            ],
+            outputs=[
+                io.Conditioning.Output(),
+                io.Float.Output(display_name="seconds"),
+            ],
+        )
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy_extras/nodes_minimax_music.py:31]: no? thats just clutter and bad design have  to obscurly pull it from conditioning. Bad rabbit.

File: comfy/model_prefetch.py [hunk 6]
```diff
@@ -17,39 +26,74 @@ def cleanup_prefetched_modules(comfy_modules):
...
+        if core is not None:
+            core()
         return
 
+    capture_stream = None
+    if enable_graph:
+        capture_stream = GRAPH_CAPTURE_STREAMS.get(device)
+        if capture_stream is None:
+            capture_stream = torch.cuda.Stream(device=device)
+            GRAPH_CAPTURE_STREAMS[device] = capture_stream
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/model_prefetch.py:61]: thats not the point of the lock and it is a pedantic complication (you are in contradiction with AGENTS.md adding paths that dont exist). the GRAPH_CAPTURE_LOCK is for outside entities that need to wait on cuda in entriity to protect the graph, it is not a multi-threaded mutex on specific graphs. The graph mutation is guaranteed single threaded.

File: comfy/text_encoders/minimax_music.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+class MiniMaxMusic3TEModel(MiniMaxMusic3AR):
+    def __init__(self, device="cpu", dtype=None, model_options={}):
+        dtype = torch.bfloat16
+        quant_config = model_options.get("quantization_metadata", None)
+        operations = model_options.get("custom_operations", None)
+        if operations is None:
+            operations = comfy.ops.mixed_precision_ops(quant_config, dtype) if quant_config is not None else comfy.ops.manual_cast
+        super().__init__(MODEL_CONFIG, dtype, device, operations)
+        self.dtypes = {dtype}
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/text_encoders/minimax_music.py:68]: so the thing is the default isnt valid here. Usually text enocoder dtypes apply to non-performance critical single step text encodings and default conservatively. There isnt a nice way to know if we are defaulting or overriden here so its a bit scopy for this late. we will go with the bad but stable behavior of forcing it under the assuption the AR it a little bit different to TE despite sharing the loader node.

File: comfy/model_prefetch.py [hunk 8]
```diff
@@ -17,39 +26,74 @@ def cleanup_prefetched_modules(comfy_modules):
...
     for queue in PREFETCH_QUEUES:
         for entry in queue:
             if entry is None or not isinstance(entry, tuple):
                 continue
             _, prefetch_state = entry
-            comfy_modules = prefetch_state[1]
+            prefetched_module, comfy_modules = prefetch_state
             if comfy_modules is not None:
-                cleanup_prefetched_modules(comfy_modules)
+                cleanup_prefetched_modules(prefetched_module, comfy_modules)
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/model_prefetch.py:41]: we no longer accept tuple by policy. The one caller that did this was changed to list. Plese check and withdraw.

**Review by comfyanonymous — APPROVED** [review]

**Review by alexisrolland — APPROVED** [review]
